File: 2015/day06/day06.hs
import System.IO
import Text.Regex.TDFA
import Data.Map (Map)
import qualified Data.Map as Map
import Data.List
import Debug.Trace
import Data.Set (Set)
import qualified Data.Set as Set

update_lights on_lights command = 
    let
        bounds = (map (read::String->Int))(getAllTextMatches(command=~"[0-9]+")::[String])
        (x0,y0) = (bounds!!0, bounds!!1)
        (x1,y1) = (bounds!!2, bounds!!3)
        region_affected = Set.fromList [(x, y)|x<-[x0..x1], y<-[y0..y1]]
    in
        if isPrefixOf "turn on" command then Set.union on_lights region_affected
        else if isPrefixOf "turn off" command then Set.difference on_lights region_affected
        else Set.difference (Set.union on_lights region_affected) (Set.intersection on_lights region_affected)

update_brightness brightness command =
    let
        bounds = (map (read::String->Int))(getAllTextMatches(command=~"[0-9]+")::[String])
        (x0,y0) = (bounds!!0, bounds!!1)
        (x1,y1) = (bounds!!2, bounds!!3)
        region_affected = [(x, y)|x<-[x0..x1], y<-[y0..y1]]
        turn_on lights region = Map.insertWith (+) region 1 lights
        turn_off lights region = if region `Map.member` lights then Map.adjust (\x-> if x>1 then x-1 else 0) region lights else lights
        toggle lights region = Map.insertWith (+) region 2 lights
    in
        if isPrefixOf "turn on" command then (trace command) foldl' turn_on brightness region_affected
        else if isPrefixOf "turn off" command then (trace command) foldl' turn_off brightness region_affected
        else (trace command) foldl' toggle brightness region_affected

main = do
    handle <- openFile "day06.txt" ReadMode
    contents <- hGetContents handle
    let
        commands = lines contents
        init_config = Set.empty
        init_brightness = []
        final_config = foldl' update_lights Set.empty (commands)
        final_brightness = foldl' update_brightness Map.empty (commands)
    print(length final_config)
    print(sum (Map.elems final_brightness))

File: 2015/day06/day06.py
import re

with open("day06.txt", "r") as f:
    instructions = f.read().split("\n")

def update_brightness(config, inst):
    print(inst)
    nums = [int(x) for x in re.findall(r"\d+", inst)]
    x0, y0 = nums[0], nums[1]
    x1, y1 = nums[2], nums[3]
    region = [(x, y) for x in range(x0, x1+1) for y in range(y0, y1+1)]
    if inst.startswith("turn on"):
        for p in region:
            if p in config.keys():
                config[p]+=1
            else:
                config[p] = 1
    elif inst.startswith("turn off"):
        for p in region:
            if p in config.keys():
                if config[p] > 1:
                    config[p] -= 1
                else:
                    config[p] = 0
    else:
        for p in region:
            if p in config.keys():
                config[p] += 2
            else:
                config[p] = 2
    return config

initial = {}
for i in instructions:
    initial = update_brightness(initial, i)
print(sum([initial[x] for x in initial.keys()]))
    
    
